File: core/runtime/python/test/utils/placed_ring_buffer_test.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import unittest

from xrpa_runtime.utils.memory_accessor import MemoryAccessor, MemoryOffset
from xrpa_runtime.utils.placed_ring_buffer import (
    PlacedRingBuffer,
    PlacedRingBufferIterator,
)


def assert_ring_buffer_equals(ring_buffer: "PlacedRingBuffer", start: int, end: int):
    mul = -1 if end < start else 1
    assert ring_buffer.count == 1 + mul * (end - start)

    for i in range(ring_buffer.count):
        assert ring_buffer.get_at(i).read_int(MemoryOffset(0)) == start + i * mul


class PlacedRingBufferTest(unittest.TestCase):
    INT_COUNT = 100
    BUFFER_SIZE = 100 * (PlacedRingBuffer.ELEMENT_HEADER_SIZE + 4)

    def setUp(self):
        mem_size = PlacedRingBuffer.get_mem_size(self.BUFFER_SIZE)
        self._bytes = bytearray(mem_size)
        self._mem = MemoryAccessor(self._bytes, 0, mem_size)
        assert not self._mem.is_null()

        self._test_ring_buffer = PlacedRingBuffer(self._mem, 0)
        self._test_ring_buffer.init(self.BUFFER_SIZE)

        self._test_ring_buffer2 = PlacedRingBuffer(self._mem, 0)

        assert self._test_ring_buffer.pool_size == self.BUFFER_SIZE
        assert self._test_ring_buffer.count == 0
        assert self._test_ring_buffer.start_id == 0
        assert self._test_ring_buffer.start_offset == 0
        assert self._test_ring_buffer.prewrap_offset == self.BUFFER_SIZE

        assert self._test_ring_buffer2.pool_size == self.BUFFER_SIZE
        assert self._test_ring_buffer2.count == 0
        assert self._test_ring_buffer2.start_id == 0
        assert self._test_ring_buffer2.start_offset == 0
        assert self._test_ring_buffer2.prewrap_offset == self.BUFFER_SIZE

    def tearDown(self):
        self._mem = None
        self._bytes = None
        self._test_ring_buffer = None
        self._test_ring_buffer2 = None

    def test_BasicOperations(self):
        # fill the buffer
        for i in range(self.INT_COUNT):
            (mem, id) = self._test_ring_buffer.push(4)

            assert not mem.is_null()
            assert id == i
            assert self._test_ring_buffer.count == i + 1

            mem.write_int(i, MemoryOffset(0))

            assert not self._test_ring_buffer.get_at(i).is_null()
            assert i == self._test_ring_buffer.get_at(i).read_int(MemoryOffset(0))
            assert i == self._test_ring_buffer2.get_at(i).read_int(MemoryOffset(0))

        assert self._test_ring_buffer.count == self.INT_COUNT
        assert self._test_ring_buffer2.count == self.INT_COUNT
        assert_ring_buffer_equals(self._test_ring_buffer, 0, self.INT_COUNT - 1)
        assert_ring_buffer_equals(self._test_ring_buffer2, 0, self.INT_COUNT - 1)
        assert self._test_ring_buffer.get_index_for_id(0) == 0
        assert self._test_ring_buffer2.get_index_for_id(0) == 0
        assert self._test_ring_buffer.get_index_for_id(1) == 1
        assert self._test_ring_buffer2.get_index_for_id(1) == 1
        assert self._test_ring_buffer.get_index_for_id(2) == 2
        assert self._test_ring_buffer2.get_index_for_id(2) == 2
        assert self._test_ring_buffer.get_index_for_id(3) == 3
        assert self._test_ring_buffer2.get_index_for_id(3) == 3

        # additional push should wrap
        (mem, id) = self._test_ring_buffer.push(4)
        mem.write_int(self.INT_COUNT, MemoryOffset(0))
        assert self._test_ring_buffer.count == self.INT_COUNT
        assert self._test_ring_buffer2.count == self.INT_COUNT
        assert_ring_buffer_equals(self._test_ring_buffer, 1, self.INT_COUNT)
        assert_ring_buffer_equals(self._test_ring_buffer2, 1, self.INT_COUNT)

        # shift oldest one out
        assert self._test_ring_buffer.shift().read_int(MemoryOffset(0)) == 1
        assert self._test_ring_buffer.count == self.INT_COUNT - 1
        assert self._test_ring_buffer2.count == self.INT_COUNT - 1
        assert_ring_buffer_equals(self._test_ring_buffer, 2, self.INT_COUNT)
        assert_ring_buffer_equals(self._test_ring_buffer2, 2, self.INT_COUNT)

        # verify that the index returned for IDs that have been removed/overwritten is 0,
        # and that the index for IDs in the buffer is correct
        assert self._test_ring_buffer.get_index_for_id(0) == 0
        assert self._test_ring_buffer2.get_index_for_id(0) == 0
        assert self._test_ring_buffer.get_index_for_id(1) == 0
        assert self._test_ring_buffer2.get_index_for_id(1) == 0
        assert self._test_ring_buffer.get_index_for_id(2) == 0
        assert self._test_ring_buffer2.get_index_for_id(2) == 0
        assert self._test_ring_buffer.get_index_for_id(3) == 1
        assert self._test_ring_buffer2.get_index_for_id(3) == 1

        assert (
            self._test_ring_buffer.get_at(
                self._test_ring_buffer.get_index_for_id(3)
            ).read_int(MemoryOffset(0))
            == 3
        )
        assert (
            self._test_ring_buffer2.get_at(
                self._test_ring_buffer2.get_index_for_id(3)
            ).read_int(MemoryOffset(0))
            == 3
        )

        # additional push should not wrap now that there is room
        (mem, id) = self._test_ring_buffer.push(4)
        mem.write_int(self.INT_COUNT + 1, MemoryOffset(0))

        assert self._test_ring_buffer.count == self.INT_COUNT
        assert self._test_ring_buffer2.count == self.INT_COUNT
        assert_ring_buffer_equals(self._test_ring_buffer, 2, self.INT_COUNT + 1)
        assert_ring_buffer_equals(self._test_ring_buffer2, 2, self.INT_COUNT + 1)

        # now shift everything out
        for i in range(self.INT_COUNT):
            assert self._test_ring_buffer.count == self.INT_COUNT - i
            assert self._test_ring_buffer2.count == self.INT_COUNT - i
            assert_ring_buffer_equals(self._test_ring_buffer, 2 + i, self.INT_COUNT + 1)
            assert_ring_buffer_equals(
                self._test_ring_buffer2, 2 + i, self.INT_COUNT + 1
            )

            assert self._test_ring_buffer.shift().read_int(MemoryOffset(0)) == 2 + i

        assert self._test_ring_buffer.count == 0
        assert self._test_ring_buffer2.count == 0

        # additional shift should return null
        assert self._test_ring_buffer.shift().is_null()
        assert self._test_ring_buffer2.shift().is_null()
        assert self._test_ring_buffer.count == 0
        assert self._test_ring_buffer2.count == 0

    def test_MixedSizes(self):
        (mem, id) = self._test_ring_buffer.push(396)
        mem.write_int(0, MemoryOffset(0))
        assert self._test_ring_buffer.count == 1
        assert self._test_ring_buffer.start_id == 0
        assert self._test_ring_buffer.prewrap_offset == 800

        (mem, id) = self._test_ring_buffer.push(196)
        mem.write_int(0, MemoryOffset(0))
        assert self._test_ring_buffer.count == 2
        assert self._test_ring_buffer.start_id == 0
        assert self._test_ring_buffer.prewrap_offset == 800

        (mem, id) = self._test_ring_buffer.push(396)
        mem.write_int(0, MemoryOffset(0))
        assert self._test_ring_buffer.count == 2
        assert self._test_ring_buffer.start_id == 1
        assert self._test_ring_buffer.prewrap_offset == 600

    def test_Iterator(self):
        iter = PlacedRingBufferIterator()
        iter.set_to_end(self._test_ring_buffer)
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is False

        # push a value in
        (mem, id) = self._test_ring_buffer.push(396)
        mem.write_int(10, MemoryOffset(0))
        assert self._test_ring_buffer.count == 1
        assert self._test_ring_buffer.start_id == 0
        assert self._test_ring_buffer.prewrap_offset == 800
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is True

        # get the value using the iterator
        mem = iter.next(self._test_ring_buffer)
        assert mem.is_null() is False
        assert mem.read_int(MemoryOffset(0)) == 10
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is False

        # push another value in
        (mem, id) = self._test_ring_buffer.push(196)
        mem.write_int(20, MemoryOffset(0))
        assert self._test_ring_buffer.count == 2
        assert self._test_ring_buffer.start_id == 0
        assert self._test_ring_buffer.prewrap_offset == 800
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is True

        # get the value using the iterator
        mem = iter.next(self._test_ring_buffer)
        assert mem.is_null() is False
        assert mem.read_int(MemoryOffset(0)) == 20
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is False

        # overflow the iterator
        for _i in range(self.INT_COUNT):
            (mem, id) = self._test_ring_buffer.push(20)
            mem.write_int(0, MemoryOffset(0))

        assert iter.has_missed_entries(self._test_ring_buffer) is True
        assert iter.has_next(self._test_ring_buffer) is True
        iter.set_to_end(self._test_ring_buffer)
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is False

        # shift everything out of the ring buffer
        while self._test_ring_buffer.count > 0:
            self._test_ring_buffer.shift()
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is False

        # push a value into the ring buffer and remove it
        (mem, id) = self._test_ring_buffer.push(396)
        mem.write_int(60, MemoryOffset(0))
        self._test_ring_buffer.shift()
        assert iter.has_missed_entries(self._test_ring_buffer) is True
        assert iter.has_next(self._test_ring_buffer) is True

        # reset the iterator so it is valid again
        iter.set_to_end(self._test_ring_buffer)
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is False

        # insert another item into the ring buffer
        (mem, id) = self._test_ring_buffer.push(396)
        mem.write_int(30, MemoryOffset(0))
        assert self._test_ring_buffer.count == 1
        assert self._test_ring_buffer.start_id == 103
        assert self._test_ring_buffer.prewrap_offset == 800
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is True

        # get the value using the iterator
        mem = iter.next(self._test_ring_buffer)
        assert mem.is_null() is False
        assert mem.read_int(MemoryOffset(0)) == 30
        assert iter.has_missed_entries(self._test_ring_buffer) is False
        assert iter.has_next(self._test_ring_buffer) is False

File: core/runtime/cs/test/utils/PlacedRingBufferTest.cs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using Xrpa;

[TestFixture]
public class PlacedRingBufferTest
{
    private static readonly int INT_COUNT = 100;
    private static readonly int BUFFER_SIZE = INT_COUNT * (PlacedRingBuffer.ELEMENT_HEADER_SIZE + 4);

    private AllocatedMemory _memory;
    private PlacedRingBuffer _testRingBuffer;
    private PlacedRingBuffer _testRingBuffer2;

    [OneTimeSetUp]
    public void SetUp()
    {
        int memSize = PlacedRingBuffer.GetMemSize(BUFFER_SIZE);
        Assert.AreEqual(memSize, 24 + BUFFER_SIZE);

        _memory = new AllocatedMemory(memSize);
    }

    [OneTimeTearDown]
    public void TearDown()
    {
        _memory.Dispose();
    }

    // order matters
    private static void AssertRingBufferEquals(PlacedRingBuffer testRingBuffer, int start, int end)
    {
        int mul = end < start ? -1 : 1;
        Assert.AreEqual(testRingBuffer.Count, 1 + mul * (end - start));
        for (int i = 0; i < testRingBuffer.Count; ++i)
        {
            Assert.AreEqual(testRingBuffer.GetAt(i).ReadInt(new MemoryOffset(0)), start + i * mul);
        }
    }

    [SetUp]
    public void Init()
    {
        _testRingBuffer = new PlacedRingBuffer(_memory.Accessor, 0);
        _testRingBuffer.Init(BUFFER_SIZE);

        _testRingBuffer2 = new PlacedRingBuffer(_memory.Accessor, 0);

        Assert.AreEqual(_testRingBuffer.PoolSize, BUFFER_SIZE);
        Assert.AreEqual(_testRingBuffer.Count, 0);
        Assert.AreEqual(_testRingBuffer.StartID, 0);
        Assert.AreEqual(_testRingBuffer.StartOffset, 0);
        Assert.AreEqual(_testRingBuffer.PrewrapOffset, BUFFER_SIZE);

        Assert.AreEqual(_testRingBuffer2.PoolSize, BUFFER_SIZE);
        Assert.AreEqual(_testRingBuffer2.Count, 0);
        Assert.AreEqual(_testRingBuffer2.StartID, 0);
        Assert.AreEqual(_testRingBuffer2.StartOffset, 0);
        Assert.AreEqual(_testRingBuffer2.PrewrapOffset, BUFFER_SIZE);
    }

    [Test]
    public void TestBasicOperations()
    {
        int id = 0;

        // fill the buffer
        for (int i = 0; i < INT_COUNT; ++i)
        {
            _testRingBuffer.Push(4, ref id).WriteInt(i, new MemoryOffset(0));
            Assert.AreEqual(_testRingBuffer.GetAt(i).ReadInt(new MemoryOffset(0)), i);
            Assert.AreEqual(_testRingBuffer2.GetAt(i).ReadInt(new MemoryOffset(0)), i);
        }

        Assert.AreEqual(_testRingBuffer.Count, INT_COUNT);
        Assert.AreEqual(_testRingBuffer2.Count, INT_COUNT);
        AssertRingBufferEquals(_testRingBuffer, 0, INT_COUNT - 1);
        AssertRingBufferEquals(_testRingBuffer2, 0, INT_COUNT - 1);
        Assert.AreEqual(_testRingBuffer.GetIndexForID(0), 0);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(0), 0);
        Assert.AreEqual(_testRingBuffer.GetIndexForID(1), 1);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(1), 1);
        Assert.AreEqual(_testRingBuffer.GetIndexForID(2), 2);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(2), 2);
        Assert.AreEqual(_testRingBuffer.GetIndexForID(3), 3);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(3), 3);

        // additional push should wrap
        _testRingBuffer.Push(4, ref id).WriteInt(INT_COUNT, new MemoryOffset(0));
        Assert.AreEqual(_testRingBuffer.Count, INT_COUNT);
        Assert.AreEqual(_testRingBuffer2.Count, INT_COUNT);
        AssertRingBufferEquals(_testRingBuffer, 1, INT_COUNT);
        AssertRingBufferEquals(_testRingBuffer2, 1, INT_COUNT);

        // shift oldest one out
        Assert.AreEqual(_testRingBuffer.Shift().ReadInt(new MemoryOffset(0)), 1);
        Assert.AreEqual(_testRingBuffer.Count, INT_COUNT - 1);
        Assert.AreEqual(_testRingBuffer2.Count, INT_COUNT - 1);
        AssertRingBufferEquals(_testRingBuffer, 2, INT_COUNT);
        AssertRingBufferEquals(_testRingBuffer2, 2, INT_COUNT);

        // verify that the index returned for IDs that have been removed/overwritten is 0,
        // and that the index for IDs in the buffer is correct
        Assert.AreEqual(_testRingBuffer.GetIndexForID(0), 0);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(0), 0);
        Assert.AreEqual(_testRingBuffer.GetIndexForID(1), 0);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(1), 0);
        Assert.AreEqual(_testRingBuffer.GetIndexForID(2), 0);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(2), 0);
        Assert.AreEqual(_testRingBuffer.GetIndexForID(3), 1);
        Assert.AreEqual(_testRingBuffer2.GetIndexForID(3), 1);

        Assert.AreEqual(_testRingBuffer.GetAt(_testRingBuffer.GetIndexForID(3)).ReadInt(new MemoryOffset(0)), 3);
        Assert.AreEqual(_testRingBuffer2.GetAt(_testRingBuffer2.GetIndexForID(3)).ReadInt(new MemoryOffset(0)), 3);

        // additional push should not wrap now that there is room
        _testRingBuffer.Push(4, ref id).WriteInt(INT_COUNT + 1, new MemoryOffset(0));

        Assert.AreEqual(_testRingBuffer.Count, INT_COUNT);
        Assert.AreEqual(_testRingBuffer2.Count, INT_COUNT);
        AssertRingBufferEquals(_testRingBuffer, 2, INT_COUNT + 1);
        AssertRingBufferEquals(_testRingBuffer2, 2, INT_COUNT + 1);

        // now shift everything out
        for (int i = 0; i < INT_COUNT; ++i)
        {
            Assert.AreEqual(_testRingBuffer.Count, INT_COUNT - i);
            Assert.AreEqual(_testRingBuffer2.Count, INT_COUNT - i);
            AssertRingBufferEquals(_testRingBuffer, 2 + i, INT_COUNT + 1);
            AssertRingBufferEquals(_testRingBuffer2, 2 + i, INT_COUNT + 1);

            Assert.AreEqual(_testRingBuffer.Shift().ReadInt(new MemoryOffset(0)), 2 + i);
        }

        Assert.AreEqual(_testRingBuffer.Count, 0);
        Assert.AreEqual(_testRingBuffer2.Count, 0);

        // additional shift should return null
        Assert.IsTrue(_testRingBuffer.Shift().IsNull());
        Assert.IsTrue(_testRingBuffer2.Shift().IsNull());
        Assert.AreEqual(_testRingBuffer.Count, 0);
        Assert.AreEqual(_testRingBuffer2.Count, 0);
    }

    [Test]
    public void TestMixedSizes()
    {
        int id = 0;

        _testRingBuffer.Push(396, ref id).WriteInt(0, new MemoryOffset(0));
        Assert.AreEqual(_testRingBuffer.Count, 1);
        Assert.AreEqual(_testRingBuffer.StartID, 0);
        Assert.AreEqual(_testRingBuffer.PrewrapOffset, 800);

        _testRingBuffer.Push(196, ref id).WriteInt(0, new MemoryOffset(0));
        Assert.AreEqual(_testRingBuffer.Count, 2);
        Assert.AreEqual(_testRingBuffer.StartID, 0);
        Assert.AreEqual(_testRingBuffer.PrewrapOffset, 800);

        _testRingBuffer.Push(396, ref id).WriteInt(0, new MemoryOffset(0));
        Assert.AreEqual(_testRingBuffer.Count, 2);
        Assert.AreEqual(_testRingBuffer.StartID, 1);
        Assert.AreEqual(_testRingBuffer.PrewrapOffset, 600);
    }

    [Test]
    public void TestIterator()
    {
        PlacedRingBufferIterator iter = new();
        int id = 0;

        iter.SetToEnd(_testRingBuffer);
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), false);

        // push a value in
        _testRingBuffer.Push(396, ref id).WriteInt(10, new MemoryOffset(0));
        Assert.AreEqual(_testRingBuffer.Count, 1);
        Assert.AreEqual(_testRingBuffer.StartID, 0);
        Assert.AreEqual(_testRingBuffer.PrewrapOffset, 800);
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), true);

        // get the value using the iterator
        {
            var mem = iter.Next(_testRingBuffer);
            Assert.AreEqual(mem.IsNull(), false);
            Assert.AreEqual(mem.ReadInt(new MemoryOffset(0)), 10);
        }
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), false);

        // push another value in
        _testRingBuffer.Push(196, ref id).WriteInt(20, new MemoryOffset(0));
        Assert.AreEqual(_testRingBuffer.Count, 2);
        Assert.AreEqual(_testRingBuffer.StartID, 0);
        Assert.AreEqual(_testRingBuffer.PrewrapOffset, 800);
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), true);

        // get the value using the iterator
        {
            var mem = iter.Next(_testRingBuffer);
            Assert.AreEqual(mem.IsNull(), false);
            Assert.AreEqual(mem.ReadInt(new MemoryOffset(0)), 20);
        }
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), false);

        // overflow the iterator
        for (int i = 0; i < INT_COUNT; ++i)
        {
            _testRingBuffer.Push(20, ref id).WriteInt(0, new MemoryOffset(0));
        }
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), true);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), true);
        iter.SetToEnd(_testRingBuffer);
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), false);

        // shift everything out of the ring buffer
        while (_testRingBuffer.Count > 0)
        {
            _testRingBuffer.Shift();
        }
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), false);

        // push a value into the ring buffer and remove it
        _testRingBuffer.Push(396, ref id).WriteInt(60, new MemoryOffset(0));
        _testRingBuffer.Shift();
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), true);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), true);

        // reset the iterator so it is valid again
        iter.SetToEnd(_testRingBuffer);
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), false);

        // insert another item into the ring buffer
        _testRingBuffer.Push(396, ref id).WriteInt(30, new MemoryOffset(0));
        Assert.AreEqual(_testRingBuffer.Count, 1);
        Assert.AreEqual(_testRingBuffer.StartID, 103);
        Assert.AreEqual(_testRingBuffer.PrewrapOffset, 800);
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), true);

        // get the value using the iterator
        {
            var mem = iter.Next(_testRingBuffer);
            Assert.AreEqual(mem.IsNull(), false);
            Assert.AreEqual(mem.ReadInt(new MemoryOffset(0)), 30);
        }
        Assert.AreEqual(iter.HasMissedEntries(_testRingBuffer), false);
        Assert.AreEqual(iter.HasNext(_testRingBuffer), false);
    }
}
